### PR TITLE
增加pipeline流程end事件的回调和网关校验无效问题修复

### LINKED ISSUE #785: pipeline条件判断有多重组合的情况下不生效
问题描述
=======
分支网关条件设置为多重组合的情况下，无法正确判断

重现方法
=======
设置如下条件， 直接返回false
(1 > 2 or 2 > 3) or ("s" > "s" or "su" > "st")

关键信息
=======
请提供以下信息:
 - [x] bk-sops   所有版本 (V3.3.X 以下)：

### LINKED ISSUE #786: 新增pipeline流程结束事件的回调功能
需求描述：
在pipeline结束之后，需要发出结束信号，做流程后续业务处理

### PR DESCRIPTION
**Pull request by selinagyan**
- 新功能
    - 增加pipeline流程结束事件信号
    close #786 

- bug修复
   - 网关的条件中配置多层级校验无效问题
    close #785

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/bk-sops?pullRequest=782) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/bk-sops?pullRequest=782) before we can accept your contribution.<br/><hr/>**selinagyan** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/bk-sops?pullRequest=782) it.</sub>

**Comment by selinagyan** [comment]
#785

**Comment by pagezhou** [comment]
关闭Issue使用关键字 
close #ID
只是关联，未完成，使用
link #ID

File: pipeline/core/flow/event.py [hunk 1]
```diff
@@ -13,6 +13,9 @@
 
 from abc import ABCMeta
 from pipeline.core.flow.base import FlowNode
+from django_signal_valve import valve
```
Comment (homholueng) [inline pipeline/core/flow/event.py]: 包导入顺序存在问题

File: pipeline/core/flow/event.py [hunk 2]
```diff
@@ -55,3 +58,4 @@ def pipeline_finish(self, root_pipeline_id):
             PipelineInstance.objects.set_finished(root_pipeline_id)
         except PipelineInstance.DoesNotExist:  # task which do not belong to any instance
             pass
+        valve.send(signals, 'pipeline_end', sender=Pipeline, root_pipeline_id=root_pipeline_id)
```
Comment (homholueng) [inline pipeline/core/flow/event.py]: 这个信号不需要经过 `valve` 来发，直接使用 signal 发即可。

File: pipeline/engine/signals/dispatch.py [hunk 3]
```diff
@@ -29,6 +40,14 @@ def dispatch_pipeline_ready():
     )
 
 
+def dispatch_pipeline_end():
+    signals.pipeline_end.connect(
```
Comment (homholueng) [inline pipeline/engine/signals/dispatch.py]: 缺少 `pipeline_end` Signal 的定义？

File: pipeline/core/flow/event.py [hunk 4]
```diff
@@ -55,3 +57,4 @@ def pipeline_finish(self, root_pipeline_id):
             PipelineInstance.objects.set_finished(root_pipeline_id)
         except PipelineInstance.DoesNotExist:  # task which do not belong to any instance
             pass
+        pipeline_end.send(sender=Pipeline, root_pipeline_id=root_pipeline_id)
```
Comment (homholueng) [inline pipeline/core/flow/event.py]: 这里需要进行异常捕获处理，不然可能会导致流程执行失败。

Comment (selinagyan) [reply pipeline/core/flow/event.py]: 感觉没必要，用户自己写的业务逻辑出现异常，不应该在这里被吃掉

Comment (homholueng) [reply pipeline/core/flow/event.py]: 这里暂时不处理，后续通过别的接口让用户定义业务逻辑。

**Pull request merged**